### ISSUE TITLE
AGS dV Calculation Error

### ISSUE THREAD

**Issue by rcflyinghokie** [virtualagc/virtualagc#1052]
Two things I have noticed when using the AGS with NASSP.  

First, address 267 does not properly calculate the total velocity to be gained.  

Second, and it could be related but I do not know, when making a burn, address 500 does read the proper velocity to be gained (even though 267 does not) however when it reaches 20fps, it zeros out.  In AGS control this initiates the engine stop command and in PGNS control, it starts counting into the negative and zeros out again.  Something to look into perhaps?

**Comment by rburkey2005** [comment]
@rcflyinghokie  I'll take a look at it, though I can't necessarily promise to have a timely solution, since it has been a dozen years or so, and I'll basically have to relearn it all from scratch.  :smile:  Any additional info you come up with would probably be very helpful.

**Comment by rcflyinghokie** [comment]
Thanks Ron!  I am playing with it more and seeing if I can find any other inconsistencies.  I will be messing with the rendezvous programs soon and will see what sort of numbers they deliver.

Anything in particular you want me to look for/test?

**Comment by indy91** [comment]
The DV calculation error is just one of many smaller and bigger issues with the Virtual AGC we have experience. I have a few to add as well. I'll try to come up with a list of things I have encounter since we started using the AEA in NASSP.

**Comment by rcflyinghokie** [comment]
I will also add what Niklas told me and I saw myself, is address 433 (LM inertial velocity) during PDI follows the PGNS perfectly until about 1000fps where is slowly diverges from the PGNS, not decreasing nearly as fast, and then jumps from about 700 to zero when the PGNS is at about 20 fps.

**Comment by rburkey2005** [comment]
@rcflyinghokie Not really, at this point, since I'm so clueless I don't know what to look for, so I'm just hoping for a miraculous clue!  :smile:

On the matter of address 267, though, I do notice that while the source code for Flight Program 6 internally lists 267 as being DELVG "VEL-TO-BE-GAINED FOR 40 MS", Flight Program 8 instead lists it as  5K18 "RD3DOT LOWER LIMIT".  Is it possible you're using FP8 rather than FP6?  Could be a documentation error in that case.

**Comment by rcflyinghokie** [comment]
> On the matter of address 267, though, I do notice that while the source code for Flight Program 6 internally lists 267 as being DELVG "VEL-TO-BE-GAINED FOR 40 MS", Flight Program 8 instead lists it as 5K18 "RD3DOT LOWER LIMIT". Is it possible you're using FP8 rather than FP6? Could be a documentation error in that case.

No this is using FP6 currently.

**Comment by thewonderidiot** [comment]
So, this may or may not be related to the problems at hand, but I've been reading through old comments about potential problem areas -- particularly those about DVP. Call me crazy, but given that the AGS is, like the AGC, armed with only a simple parallel adder, also performs division as a DP dividend and SP divisor, _and_ has the exact same requirement that the magnitude of the divisor be greater than that of the dividend... I think there is a very solid chance that the two used the same basic algorithm for performing division. That algorithm, in simple terms, is:

1) Calculate and store the negative absolute value of the DP dividend.
2) Shift the dividend up one.
3) Add the absolute value of the divisor to the upper word of the dividend. If the dividend becomes positive, discard the sum and take a 0 as the next quotient bit. If the dividend remains negative (or becomes zero), retain the sum as the new dividend and take a 1 as the next quotient bit.
4) Go to step 2, until all 17 data bits of the quotient have been calculated.
5) Examine dividend and divisor signs to apply the correct sign to the quotient.

From the AGC point of view, the DVP at address 7701 is a "nonsense" division, but following the AGC algorithm above provides the expected 177777 result, assuming I did my math correctly. (Plus, step 3 will generate 1s forever with the 7701 numbers, which means it would round to 200000 if I'm reading this correctly.)

Given that the programming manual seems to discourage these cases, it's unlikely that the numeric results actually matter for anything...

**Comment by thewonderidiot** [comment]
Actually, scratch that -- apparently my brain defaults to one's complement now. The above algorithm only gets the expected results for the 7701 DVP if the running dividend sum is not allowed to hit 0. In other words, if the divisor divides evenly into the dividend. In all such cases, then (dividend % divisor == 0), the result would be (answer - 1).77777777....

That seems like a pretty serious (and weird, and artificial) limitation, but considering they also give the rounded result, the rounded result would always be correct. But if they use the unrounded result, it could lead to somewhat frequent errors.

**Comment by thewonderidiot** [comment]
Throwing out a couple more possibilities from looking through the instruction set -- ABS and COM both say that both 0 and "minus one" are unchanged by the operation. The manual doesn't say so, but I wonder if it's possible that the overflow indicator is set in the "minus one" case? If they don't explicitly inhibit overflow during those instructions, I could very well see the two's complement addition of 1 being interpreted as an overflow.

**Comment by rburkey2005** [comment]
I've looked at the problem a bit, and haven't found anything suspicious yet.

@thewonderidiot Your theory sounds plausible, but any idea as to how to check it?

**Comment by thewonderidiot** [comment]
I'm working on it.... in a very round-about way, and it's going to take a bit of time to get there. Since it would be really slow to have RC or Niklas test every possible theory (and probably annoying for them), I've installed Orbiter and NASSP and am working on learning to use it and fly the LM myself, so I can do quick turnarounds on attempted fixes like these. But like I said, it's going to take a bit :smile:

I'm also going through everything we pulled down from the archived NTRS to see if I have anything (so far I've come up with a few Grumman requirements/interface documents but nothing that dives into the AEA architecture itself), and I'm also checking our regular sources. UHCL has a large study guide about the abort guidance section that looks promising.

**Comment by thewonderidiot** [comment]
So, I think I'm missing or misunderstanding something here. Apparently the Guidance Mode word (aka S10, at address 410), has six valid settings, from 0 to 6. Niklas says these are:
0: Orbit Insertion 
1: CSI
2: CDU
3: TPI Search
4: TPI Execute
5: External DV

But the tests on S10 don't make any sense at all to me. For example:
```
#
#	GUIDANCE MODE CHECK
#
GM	CLA	S10
	ADD	3B3
	TMI	XDV		# EXTERNAL DELTA V
```
3B3 is an instruction in the self-test used as a value, here. The listing says it assembles thusly:
```
10703: 7520 14 0 0000              	3B3   	LDQ     TS0             # 400000   525252   525252
```

So, it has a value of 140000, if I am understand this correctly. So for their math here, they are loading a value which can range from 0 to 5 into the accumulator. They then add 140000 to it. And if the result becomes negative then the value must have been a 5? This doesn't make any sense to me, for any location of the 5 within the word (so even, say, 500000 or 050000 doesn't seem to work). What is happening here?

**Comment by thewonderidiot** [comment]
Hm, nevermind, I guess it is set in those upper bits. The operating instructions say that to use External DV, set 410 to +50000.

**Comment by thewonderidiot** [comment]
....although, the math still doesn't make sense. 140000 + 50000 = 210000, which is still positive...

**Comment by thewonderidiot** [comment]
Hah, I was hoping you knew! I saw one of those asterisks a few days back and spent a while trying to figure out what the heck it meant, to no avail. The best I could come up with was that perhaps it was similar to the asterisks in YUL/GAP, indicating a change.

**Comment by thewonderidiot** [comment]
So, looking through FP6 and FP8 to see if there is any pattern among those asterisks, I found something interesting. At the bottom of page 110 of FP8, somebody has manually drawn in an asterisk next to an AXT 4,1.

**Comment by thewonderidiot** [comment]
Okay, possible theory, although the above AXT doesn't really fit the bill. It sort of looks like * is used to denote instructions that _might_ cause an overflow. At least, they're very frequently followed by comments talking about testing for overflow.

**Comment by thewonderidiot** [comment]
Okay, that seems to be reasonably consistent -- I'm going to go with that explanation for now. There's even one page in FP8 where an asterisk was written next to a potential overflow-causing instruction, and the TOV a bit down the page was underlined. (I also noticed that a lot of constants have asterisks next to OCT or DEC, which obviously does not fit this pattern).

Assuming this is the case, though, there are instances of both *COM and *ABS in the programs, and with *ABS in particular it looks like there may be a handful of areas where this could affect program flow. So, @rcflyinghokie, the first set of patches I'm going to give you will make both COM and ABS set overflow if the accumulator is negmax. If that doesn't work, then we'll try it with only ABS setting overflow. I'll pull request the potential fixes to dseagrav/NASSP for now, since I think it'll be easier for you to integrate them that way. I'll start fiddling with DVP implementations in the meantime.

**Comment by thewonderidiot** [comment]
I have pull requested these changes in the PR linked above. The self-tests still pass with these changes, so nothing fundamental is broken, but I can't promise the same for mission software. :smile:

**Comment by rcflyinghokie** [comment]
@thewonderidiot I just put those changes into my testing branch and will give them a try!

**Comment by rburkey2005** [comment]
@thewonderidiot  Regarding the code that tests for the guidance mode, it appears to me that what's loaded into S10 is not 50000 or 500000, but rather 240000; i.e., the 5 skips the sign bit, and then is loaded into the 3 most-significant non-sign bits.  So if this value adds to 140000, you get 400000, and so it has flipped to minus.  I get this impression by looking at table 5.3 in http://www.ibiblio.org/apollo/Pultorak_files/FP6_OperatingManual.pdf, which is actually discussion a of EONS10 (engine-on bit plus S10) rather than S10 itself. See also the entry for E10/S10 on p. 5-6 of that doc.  (Though admittedly, I'm in a bit of a rush to get out the door and haven't fully explored this idea just yet, and am even more likely to be making some daft error than I usually am.)

**Comment by thewonderidiot** [comment]
Ah, nice, that makes perfect sense, thanks!

(At least, mathematical sense. So the DEDA is showing all but the least significant 2 bits? How strange.)

**Comment by thewonderidiot** [comment]
So I tested out a crackpot theory earlier, and I'm almost completely positive that it's wrong, but it's worth recording here just in case it turns out to have any merit:

The erasable memory in the AGS is a very traditional 4-wire coincident-current core memory -- it has X and Y selection lines, a sense line, and an inhibit line. The permanent memory is sort of a hack on that. It removes the inhibit line, and the X selection line is wired such that it only runs through cores which represent zeros.

This means that only the cores representing 0s can be switched, and so there are two immediate implications:
1. The cores "storing" 1s are really just taking up volume and mass.
2. Because the "1" cores can't be switched, a pulse detected on the sense line must necessarily mean that the core that was switched was storing a 0. The AGS therefore used the opposite polarity for storage as the AGC (for the AGC, a detected pulse meant a 1).

What's more, this means that readout of the fixed memory is actually destructive (!). After a read cycle, all of the 0 cores will have been flipped and so would now read as a 1, so unless the writeback cycle is performed, you would get all 1s the next time you read the address. Since the inhibit lines are removed for this portion of the memory, though, it's only possible to write 0s, and only to the cores that are supposed to be 0. So it doesn't matter what is written back out, just that _something_ is.

Anyways, presumably this implies that the data stored in the erasable half has the same polarity as the fixed half (pulse on sense line = 0, no pulse = 1). Since the documentation suggests using CLZ, ADZ, etc. to save power by _not_ writing values back out, there's an interesting possibility for these instructions. I had been thinking that they skipped the write cycle entirely -- but since reading a location sets it to all 1s instead of all 0s, as in the AGC, this would mean that the locations would not be set to 0 but rather to -1. For the locations to get zeroed, they would need to still perform the writeback cycle, but inhibit activity on the inhibit lines. (This would still save up to 90% of the power, though, depending on how many 1s were going to be written).

So, I tested out having all of the **Z instructions set the accessed location to 777777 instead of 000000. Unsurprisingly, I guess, the READ OUT button ceased to work, although address entry and the CLR button were fine. So I never was able to even check if the self tests still passed. :smile:

**Comment by thewonderidiot** [comment]
Along more actually useful lines, @rcflyinghokie gave me this NASSP scenario. The LM is in lunar orbit, and the AGS has been powered on and given correct state vectors. 

[Apollo 11 AGS Testing.scn.zip](https://github.com/virtualagc/virtualagc/files/1373574/Apollo.11.AGS.Testing.scn.zip)

Those state vectors are propagating correctly. Some derived products are not, however. The addresses that we definitely know are misbehaving are:
* DELVG [267] -- velocity-to-be-gained for 40ms
* QA [315] -- LM apocynthion
* QLTELE [403] -- LM pericynthion
* DELVGX, DELVGY, DELVGZ [500-502] -- thrust velocity to gain

Of these, it looks like the apsides might be the easiest to study. QA is calculated as
```
QA = (<some number> * 2) - QLTELE
```

QLTELE is derived directly from QL [343], which is calculated in LMORB (they differ by the landing site radius). So, QL is a very good candidate for study. Unfortunately the location is shared with EONS10. So earlier I gave RC a build of FP6 that skips use of the location as EONS10, instead only using it for pericynthion. He observed that the location is normally 0, but changes to -83886 during burns. It does look like LMORB can set QL to a constant (the value in address 216) if an overflow is encountered, but we haven't yet tested to see if that is what's happening.

**Comment by thewonderidiot** [comment]
@rcflyinghokie recorded these videos of AGS anomalies for us, showing debug lines with various misbehaving memory addresses displayed in native octal:
* https://www.dropbox.com/s/b3jnwfqzw5me7zj/Surface%20Oddity.flv?dl=0 -- 500-502 doing really weird things on the surface
* https://www.dropbox.com/s/dhy9lp661o7nh1i/DOI.flv?dl=0 -- DOI burn. 500-502 track closely until ~20fps to go
* https://www.dropbox.com/s/rwh9vld0wg7mqev/PDI.flv?dl=0 -- PDI burn

470-472, where the velocity is accumulated, appear to be working correctly.

**Comment by thewonderidiot** [comment]
Today, @rcflyinghokie, @indy91, and I spent some time looking over and discussing the apsides calculations. We determined that the calculations for QL [343], which I talked about above, are fairly simple. It is fully determined from only four values: HDOT [367], V [433], VH [427], and R [210]. Furthermore, the AGS appears to be calculating incorrect QL results during times when all four of these precursors have correct values. (!)

Niklas and Ryan gave me the following sample known good octal values for these four locations:

| Name | Addr | Value |
| -------- | ------- | ------- |
| HDOT | 367 | 777310 |
| V | 433 | 255674 |
| VH | 427 | 255674 |
| R | 210 | 257143 |

In theory, then, tracing down an issue is a simple matter of jamming these four values into the AEA with a debugger and then stepping through the LMORB routine, looking for any anomalous results:
```
EDIT PC 5444
EDIT 367 777310
EDIT 433 255674
EDIT 427 255674
EDIT 210 257143
STEP
...
```

**Comment by thewonderidiot** [comment]
And voila, I found a bug! :smile:
```
Stopped because step count reached.
A=0054741       Q=0260644       Overflow=0      Index=000       Halted=0
Icount=2        Cycles=1150
05463  0520002  LLS     00002
> step

Stopped because step count reached.
A=0263606       Q=0303220       Overflow=1      Index=000       Halted=0
Icount=11       Cycles=1169
05464  0100010  STO     00010
>
```
`LLS 2` should definitely _not_ be setting the overflow flag here. And an overflow flag at any point in these computations makes LMORB give up and fall back on the default QL value. The problem lies in this line of code:
```
	  llk = (0177777777777LL & ~((1 << (34 - i)) - 1));
```
The `(1 << (34-i))` is treated as a 32-bit integer, and so overflows, making the overall mask calculation incorrect. The `1` needs to be changed to a `1LL` to work. Interestingly enough, the equivalent line in LRS does have the `LL`:
```
        llk = (0177777777777LL & ~((1LL << (34 - i)) - 1));
```

With this correction in place, LMORB is able to make it all the way through the calculations without getting an overflow, and determines QL to be 257116. @rcflyinghokie, I'll add this change to dseagrav/NASSP#168.

**Comment by rburkey2005** [comment]
@thewonderidiot I don't believe I personally added any pthread_cond_XXXX code.  I think that Jordan Slott may have added it when he implemented using readline (with all of its fancy editing) to get the input line from the keyboard.  However, readline ended up being a cross-platform nightmare, and so we eventually permanently disabled it.  The disabling may have been done imperfectly in yaAGS, since the AGS stuff was new at the time and probably nobody was using it, and the problem with debugging might not have been noticed.  The corresponding pthread_cond_broadcast in yaAGC is conditioned on whether or not readline is being used, and in effect is not commented out.

**Comment by rburkey2005** [comment]
@thewonderidiot Oh, and fantastic work finding that bug!

**Comment by rcflyinghokie** [comment]
@thewonderidiot I will throw that in my testing branch and give it a whirl!  Does it matter which binary of FP6 I try it with?

**Comment by thewonderidiot** [comment]
Aha, thanks Ron! I've made that same change to yaAGS.

Ryan hasn't quite completed his rendezvous yet, but I think all of the major burns happened without error, and in very close agreement with the AGC. I think there are now no known AGS emulation problems. Ryan's been testing with overflow tweaks to COM and ABS, so we're not yet sure if these are actually needed.

**Comment by rcflyinghokie** [comment]
Looks like all phases of rendezvous work as advertised in lunar orbit.  I used AGS calculations for every part and ags controlled burns as well with RR and PGNS as backup/basis of comparison.  The TPI numbers were the only ones that differed from the PGNS with any significance but regardless it got me back to the CSM.  I am certain a lot was user error as the CSI/CDH are very straight forward but the TPI targeting routine I am still learning exactly how it is used.

Long story short, bravo zulu @thewonderidiot  for finding the apparent "fix of all fixes" in the AGS!

**Comment by rburkey2005** [comment]
@rcflyinghokie @thewonderidiot @indy91 I'm really glad to see the AGS being checked out in NASSP.  It's probably a silly question to ask here &mdash; and an ignorant one in general &mdash; but anyway ... there's an episode of the HBO series From the Earth to the Moon in which a scene just happens to take place in a classroom training session about the AGS.  It's a dramatic, fictionalized scene, and so it isn't actually intended to convey any info about the AGS, and doesn't attempt to do so.  But in this scene, an astronaut (portraying, I believe, Edgar Mitchell) has a throway line in which he asserts, I think, that they had been "playing around in the simulator" and had concluded that the abort system might actually be usable for making the landing itself rather than simply performing an abort.  I wonder if this is really possible?  I'd certainly like to know.  And if so, I imagine it would be a cool NASSP scenario.

**Comment by rburkey2005** [comment]
@thewonderidiot Mike, since you've ended up doing all the work on this, I'm reassigning it from me to you, and you can decide if/when to close the issue.  Thanks!

**Comment by thewonderidiot** [comment]
@rcflyinghokie did one final check this morning, running through PDI, aborts, external DV, a CSI, and insertion from the surface, with test code I put in place that would notify us if ABS or COM setting the overflow flag was ever responsible for a TOV branch taken. No such cases were ever detected, so it seems safe to omit them for now. If any other problems crop up, though, I think we should bring them back and see if they help.

**Comment by indy91** [comment]
@rburkey2005 there are two possible answers for that. He could have meant the Abort Guidance System as a whole. A lot of it is done in analog circuits, while the AGC is more fly-by-wire. So the whole control scheme of the AGS isn't even in the computer. The AEA simply outputs an attitude error and the ATCA and other subsystems do a lot of the work. So the AGS is a perfectly fine system to control the LM and if you try to manually land the LM (without help from any computer) then controlling the spacecraft is no problem.

The question is if he meant a computer assisted landing with the AGS. The AGS Flight Programs have no program that could really do this. It might be possible to change a whole lot of mission constants in the AEA and use the orbital insertion program to help with a landing attempt. That would be the only routine in the AGS with a guidance law that might be helpful. I don't know the AGS guidance equations good enough yet to know if that is even remotely possible. I'll look into it at some point. Would be fun to try.

**Comment by rburkey2005** [comment]
@indy91 Thanks, please do!  It's hard to tell whether a semi-fictionalized character in a drama meant anything at all, or what he meant, but it would be cool if he had really meant something.  :smile:  Unfortunately, Mitchell himself passed away last year, so we'll never be able to ask him personally about it.
